nowcasting_dataset/dataset/datasets.py

```diff
@@ -187,6 +206,79 @@ def __getitem__(self, batch_idx: int) -> example.Example:
+                    :, sat_mask, :, :, :
+                ]  # Sat is in [B, L, W, H, C]
+                _LOG.debug(
+                    f"Sat Datetime Shape: {batch['sat_datetime_index'].shape} Sat Data Shape: {batch['sat_data'].shape}"
+                )
+
+            # Now for NWP, if used
+            if "nwp" in self.required_keys and "nwp_target_time" in batch:
+                nwp_mask = np.logical_and(
+                    start_time <= batch["nwp_target_time"][0],
+                    batch["nwp_target_time"][0] <= end_time,
+                )
+                batch["nwp_target_time"] = batch["nwp_target_time"][:, nwp_mask]
+                batch["nwp"] = batch["nwp"][:, :, nwp_mask, :, :]  # NWP is in [B, C, L, W, H]
+                _LOG.debug(
+                    f"NWP Datetime Shape: {batch['nwp_target_time'].shape} NWP Data Shape: {batch['nwp'].shape}"
+                )
+
+            # Do the for time constants, as either NWP or Sat data should exist and have masks
+            for k in list(nowcasting_dataset.consts.DATETIME_FEATURE_NAMES):
+                batch[k] = batch[k][:, sat_mask if "sat_data" in self.required_keys else nwp_mask]
+
+            # Now GSP, if used
+            if GSP_YIELD in self.required_keys and GSP_DATETIME_INDEX in batch:
+                gsp_mask = np.logical_and(
+                    start_time <= batch[GSP_DATETIME_INDEX][0],
+                    batch[GSP_DATETIME_INDEX][0] <= end_time,
+                )
+                batch[GSP_DATETIME_INDEX] = batch[GSP_DATETIME_INDEX][:, gsp_mask]
+                batch[GSP_YIELD] = batch[GSP_YIELD][:, gsp_mask]  # GSP is in [B, L, N]
+                _LOG.debug(
+                    f"GSP Datetime Shape: {batch[GSP_DATETIME_INDEX].shape} GSP Data Shape: {batch[GSP_YIELD].shape}"
+                )
+
+            # Now PV systems, if used
+            if "pv_yield" in self.required_keys and "pv_datetime_index" in batch:
+                pv_mask = np.logical_and(
+                    start_time <= batch["pv_datetime_index"][0],
+                    batch["pv_datetime_index"][0] <= end_time,
+                )
+                batch["pv_datetime_index"] = batch["pv_datetime_index"][:, pv_mask]
+                batch["pv_yield"] = batch["pv_yield"][:, pv_mask]  # PV is in [B, L, N]
+                batch["pv_azimuth_angle"] = batch["pv_azimuth_angle"][
+                    :, pv_mask
+                ]  # PV is in [B, L, N]
+                batch["pv_elevation_angle"] = batch["pv_elevation_angle"][
+                    :, pv_mask
+                ]  # PV is in [B, L, N]
+                _LOG.debug(
+                    f"PV Datetime Shape: {batch['pv_datetime_index'].shape} PV Data Shape: {batch['pv_yield'].shape}"
+                    f" PV Azimuth Shape: {batch['pv_azimuth_angle'].shape} PV Elevation Shape: {batch['pv_elevation_angle'].shape}"
+                )
+
+            # TODO Add others for subsetting
```

Comment:
Do you know what other ones there are to do here?

Reply:
I'm not sure, I think I got them all, GSP, PV, NWP, Satellite, and datetime features, not sure if there are more to add right now

Reply:
If so, perhaps remove this comment, just to keep it tidy